### DISCUSSION TITLE
[Proposal] Automatic early returns on failed results

### DISCUSSION THREAD

**Discussion by michaldivis** [dotnet/csharplang#7488] [Language Ideas]
# Automatic early returns on failed results

- [x] Proposed
- [ ] Prototype: Not Started
- [ ] Implementation: Not Started
- [ ] Specification: Not Started

## Summary

A new `ensure` keyword that when put before a method call will either continue execution or return a value based on a success condition. 

Inspired by [Rust's question mark operator](https://doc.rust-lang.org/std/result/#the-question-mark-operator-).

## Motivation

The motivation is to simplify working with Result objects, specifically propagating a failed result through multiple methods.

The current way of doing that might look like this:

```csharp
public Result CreateUser(CreateUserRequest request)
{
    var normalizeUsernameResult = NormalizeUsername(request.Username);

    if (!normalizeUsernameResult.IsSuccess)
    {
        return normalizeUsernameResult;
    }
    
    var normalizeAddressResult = NormalizeAddress(request.Address);

    if (!normalizeAddressResult.IsSuccess)
    {
        return normalizeAddressResult;
    }
    
    var user = new User
    {
        Username = normalizeUsernameResult.Value,
        Address = normalizeAddressResult.Value,
    };

    var saveUserResult = SaveUser(user);    

    if (!saveUserResult.IsSuccess)
    {
        return saveUserResult;
    }

    return Result.Ok();
}

public Result<string> NormalizeUsername(string username)
{
    // implementation omitted for brevity
}

public Result<string> NormalizeAddress(string address)
{
    // implementation omitted for brevity
}

public Result SaveUser(User user)
{
    // implementation omitted for brevity
}
```

And here's a modified version that takes advantage of the proposed feature:

```csharp
public Result CreateUser(CreateUserRequest request)
{
    var normalizeUsernameResult = ensure NormalizeUsername(request.Username);
    
    var normalizeAddressResult = ensure NormalizeAddress(request.Address);
    
    var user = new User
    {
        Username = normalizeUsernameResult.Value,
        Address = normalizeAddressResult.Value,
    };

    ensure SaveUser(user);

    return Result.Ok();
}

public Result<string> NormalizeUsername(string username)
{
    // implementation omitted for brevity
}

public Result<string> NormalizeAddress(string address)
{
    // implementation omitted for brevity
}

public Result SaveUser(User user)
{
    // implementation omitted for brevity
}
```

## Detailed Design

The Rust feature this proposal is inspired by uses a `?` operator to make this feature work, however, the `?` operator is already used for something else in C#, so I propose to create a new keyword: `ensure`.

Then the following code:
```csharp
var normalizeUsernameResult = ensure NormalizeUsername(request.Username);
```

Could be lowered into:
```csharp
var normalizeUsernameResult = NormalizeUsername(request.Username);

if (!normalizeUsernameResult.IsSuccess)
{
    return normalizeUsernameResult;
}
```

For this feature to work, the following conditions will have to be met:
1. The new keyword will have to be able to determine if a return value is successful or not. To do that, I propose two approaches:
    - An interface: The returned value would have to implement a specific interface that defines the `IsSuccess` property. 
    - Ducktyping: The returned value would have to have a `IsSuccess` property, but wouldn't need to implement any specific interface.
2. The return type of the method will have to match the return type of the parent method, or be implicitly convertable to it.

**Comment by HaloFour** [comment]
I think a discussion on general comprehensions over monads is worth discussing after C# implements [discriminated unions](https://github.com/dotnet/csharplang/issues/113).

**Reply by TonyValenti** [reply]
Seven years ago I would have guessed we were right in the cusp of DUs.

**Reply by DavidArno** [reply]
In 2012, it felt like we were on the cusp of records being added in C# 5. Eight years later, they landed.

So maybe next year we'll get DUs? 😊

**Comment by jnm2** [comment]
I love this keyword. I've been wishing for syntax sugar for this for many years, and it would be great if the pattern could be flexible enough to desugar APIs like this one as well:

```cs
Result<Bar> M(int a, int b)
{
    Result<Foo> fooA = GetFoo(a);
    if (fooA.IsError(out var message))
        return Result.Error(message);
    
    Result<Foo> fooB = GetFoo(b);
    if (fooB.IsError(out message))
        return Result.Error(message);
        
    return Result.Success(new Bar(fooA.Value, fooB.Value));
}
```

To:
```cs
Result<Bar> M(int a, int b)
{
    return new Bar(ensure GetFoo(a), ensure GetFoo(b));
}
```

**Reply by michaldivis** [reply]
Wow. This one is on the extreme side, but why not?

**Comment by neon-sunset** [comment]
Much like `await`, `ensure File.TryReadAllBytes(path);` could look recognizable, and then duck-typing fits existing conventions on how features are added.

Not sure about the keyword itself...perhaps `try expr`? e.g.:
```cs
public async Task<Result<string, Error>> Example()
{
    // Standard
    var text = try File.ReadAllText("example.txt");
    
    // Async + Try
    var content = new StringContent(text);
    var response = try await _http.PostAsync("https://example.org", content);
    
    return await response.Content.ReadAsStringAsync();
}
```

**Reply by HaloFour** [reply]
Do you mean `Task<Result<string, Error>>`?  Honestly, I'm not sure you'd want to mix the two, given `Task<T>` is already like an async version of `Result<...>` and has its own way of indicating failure.  You also wouldn't want to incur the cost of async machinery overhead for purely synchronous code, which `Result<...>` would likely represent.  But it does illustrate that we already have a couple of these monad-ish types in the ecosystem, and very specific language support for coroutines or comprehensions over them which can't interoperate with each other without additional language support.  That's why my opinion is that if DUs come to pass and we can expect additional monad types to join the ecosystem that it's worth looking into ways of expanding comprehensions to be a bit more general purpose.  I don't imagine we'd have a "blessed" `Result<...>` type until DUs anyway.

**Reply by neon-sunset** [reply]
> Do you mean Task<Result<string, Error>>?

Yes, thanks.

> You also wouldn't want to incur the cost of async machinery overhead for purely synchronous code, which Result<...> would likely represent.

What kind of overhead do you have in mind if `try {expr}` is lowered to `var res = {expr}; if (!res.IsSuccess) return res`?
(or even `if (expr.IsSuccess) { /* rest of the method body */ } else return expr;` for branch ordering reasons)

The way I'm reading this proposal is for the pattern to be duck-typed so that it can exist without tying itself to the introduction of `DU`s in the future, which then will be able to take advantage of the `ensure` (or `try`) syntax, without forcing to have a blessed union type (which IIRC was the case in Rust until it introduced a trait for user-defined types to be able to use `?` too).

**Reply by HaloFour** [reply]
> The way I'm reading this proposal is for the pattern to be duck-typed so that it can exist without tying itself to the introduction of `DU`s in the future, which then will be able to take advantage of the `ensure` (or `try`) syntax,

It requires that the team settle on what that pattern would be, before the conversation of how said DUs would work.  And it also ties the language feature to a guess of how `Result<...>` specifically would work, without considering whether that pattern makes sense for other DUs like `Option<T>` or `Reader<T>` or `IO<T>` or anything else, where `IsSuccess` wouldn't make any sense.

**Reply by jnm2** [reply]
The feature could be developed with attributes specifying what properties or methods the compiler would use to put this together. We were discussing this a few times in Gitter and later, Discord.

>  Honestly, I'm not sure you'd want to mix the two, given `Task<T>` is already like an async version of `Result<...>` and has its own way of indicating failure. 

`Task<Result<X>>` is something I do frequently. `Result<X>` is there for the same reason that I'd use it in synchronous code (because I'm not modeling exceptional behavior). `Task<T>`'s way of indicating failure is to use exceptions.

**Comment by IS4Code** [comment]
Instead of/in addition to `IsSuccess`, I would make `ensure` check implicit conversion to `bool`, maybe `bool?` too.

**Reply by AartBluestoke** [reply]
the problem with cast to bool is : do you short circuit return the result object on true or false?  
code that lowers to `if(!result)  return result` feels funny

**Reply by IS4Code** [reply]
To me that feels exactly like the one with `.IsSuccess` check. Perhaps `ensure not` could be added as the negation.

**Comment by Vixenka** [comment]
Looks like a proposal is going to use text to express the contents of exceptions, that is a very big setback.

Additionally adding in `Result<T>` container of exception or string ensures using a heap, which means performance will be very similar to exceptions. And maybe more expensive (depends on the implementation of exceptions). And and maybe implementation of exceptions could be faster than many element enum container for errors which will be stores information about him.

So to gain a performance. It should only stores bool which indicates about success. But we now have a great mechanism for that:
```cs
TValue? TryGetValue(TKey key) {
    if (innerDictionary.TryGetValue(key, out TValue v))
        return v;
    return null;
}
```

But let's consider that it's not about performance and let's check for usage. Rust creators want to create transparent errors in their language. Therefore, there is no one type for errors, duly each function should create its own containing only possible exceptions to receive from this function. But time shows programmers faces and people starts creating very very big error types which starts covered not one function, but a whole project or few. Simultaneously Rust users starts to frequent use unwrap methods which and in this way it is no different from exceptions.

In my opinion we would gain much more by fine-tuning the exceptions a bit. Instead of adding `Result<T>` we can add attribute `MustCatch(params Type[])` which when executor not implement catching of exception types submitted in attribute will be causing a compile warning (not error for backwards compatibility). Additionally developer can bypass catching exception by adding it as attribute to executor method.

**Reply by michaldivis** [reply]
The `Result` class isn't a part of the proposal.

**Reply by HaloFour** [reply]
> Additionally adding in `Result<T>` container of exception or string ensures using a heap, which means performance will be very similar to exceptions.

The `Result<T>` could be a value type and live on the stack, thus no additional heap allocations required.  Also, such a type would add very little additional overhead.  If it wrapped an exception, then you would incur the cost of that exception, but not much more.  If the error case was not an exception but only a message or similar then this solution would be much, much more performant than throwing exceptions.

> Instead of adding Result<T> we can add attribute MustCatch(params Type[]) which when executor not implement catching of exception types submitted in attribute will be causing a compile warning (not error for backwards compatibility). Additionally developer can bypass catching exception by adding it as attribute to executor method.

That would be checked exceptions, and no thank you.

https://www.artima.com/articles/the-trouble-with-checked-exceptions

**Comment by HaloFour** [comment]
Two additional thoughts.

First, without a "blessed" `Result<T, E>` type like Rust I think support for a feature like this would require similar machinery to `async`/`await`, where the compiler emits code to unwrap the result on every "ensured" expression so that in an error case it can rewrap the error back into the expected result type.  That is how C# handles mixing `Task`, `Task<T>`, `ValueTask`, `ValueTask<T>` and any other task-like type today.  Without that, you'll end up in a situation with incompatible result types.  This is already the case between `Result<string, Exception>` and `Result<string, string>`, in that they are completely incompatible types and implicit conversions won't be able to cover the infinite potential permutations, especially since they can't be generic.  IMO, this is a good thing as I'd prefer that such comprehensions be heterogeneous anyway, supporting multiple flavors of monads.

Second, beyond converting between result types, you'll need to be able to convert between error types too.  Rust does this by unwrapping the error and converting via `From::from`, but I don't think that this will work here.  I think it's highly unlikely that the runtime team would ever consider implicit conversions between `Exception` and `string`, for example, and covering all of the potential error states within a single error DU would be quite a lot of boilerplate.

**Reply by michaldivis** [reply]
I definitely agree that dealing with the type conversions is very difficult and tricky.

**Comment by Joe4evr** [comment]
This reminds me that someone already made a prototype like this by (ab)using custom Task-likes years ago: #3403

I'm personally still not a fan of this paradigm, but beyond that it's good to take in the previous discussion.
